applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py

```diff
@@ -0,0 +1,716 @@
+from __future__ import print_function, absolute_import, division  # makes KratosMultiphysics backward compatible with python 2.6 and 2.7
+
+# Importing the Kratos Library
+import KratosMultiphysics
+
+# Check that applications were imported in the main script
+KratosMultiphysics.CheckRegisteredApplications("ConvectionDiffusionApplication")
+
+# Import applications
+import KratosMultiphysics.ConvectionDiffusionApplication as ConvectionDiffusionApplication
+
+# Importing the base class
+from python_solver import PythonSolver
+
+# Other imports
+import os
+
+
+def CreateSolver(model, custom_settings):
+    return ConvectionDiffusionBaseSolver(model, custom_settings)
+
+class ConvectionDiffusionBaseSolver(PythonSolver):
+    """The base class for convection-diffusion solvers.
+
+    This class provides functions for importing and exporting models,
+    adding nodal variables and dofs and solving each solution step.
+
+    Derived classes must override the function _create_solution_scheme which
+    constructs and returns a solution scheme. Depending on the type of
+    solver, derived classes may also need to override the following functions:
+
+    _create_solution_scheme
+    _create_convergence_criterion
+    _create_linear_solver
+    _create_builder_and_solver
+    _create_convection_diffusion_solution_strategy
+
+    The convection_diffusion_solution_strategy, builder_and_solver, etc. should alway be retrieved
+    using the getter functions get_convection_diffusion_solution_strategy, get_builder_and_solver,
+    etc. from this base class.
+
+    Only the member variables listed below should be accessed directly.
+
+    Public member variables:
+    model -- the model containing the modelpart used to construct the solver.
+    settings -- Kratos parameters containing solver settings.
+    """
+    def __init__(self, model, custom_settings):
+        super(ConvectionDiffusionBaseSolver, self).__init__(model, custom_settings)
+
+        default_settings = KratosMultiphysics.Parameters("""
+        {
+            "model_part_name" : "Thermic",
+            "domain_size" : -1,
+            "echo_level": 0,
+            "buffer_size": 2,
```

Comment:
Why is the buffer size a json option? I think this should not be settable by the user, since the choice of time integration scheme determines it.

Reply:
You are choosing the scheme already in the solvers, I would set the buffer size within the solver, in relation to the scheme choice (that is what we do in the fluid, at least)

Reply:
In structures we set manually, I was suggesting to add to scheme.h a method GetRequiredBuffer() or something like it, in order to not add manually each time the buffer required (needs to be defined on the scheme, or error can be triggered). Certainly I think this PR is already huge, so this can be done in a further one

Reply:
I would not add it to scheme.h, since it does not always know the buffer size (symbolic elements that integrate in time "internally" are an example of this). The python solver itself has all the information needed to determine the buffer size (scheme+element name), so it should not give the user a chance to set it to the wrong value...

Reply:
Any hint which size should I set?, I don't have that much experience with the Conv-Diff application

Reply:
I'd say 2 for dynamic problems (since this is enough both for Backward Euler and Crank-Nicolson). Probably 1 should be enough for static ones, but I admit I have never tried it.

Reply:
Well can be check easily, I added two tests, one static and other dynamic